torch/csrc/jit/OVERVIEW.md

```diff
@@ -886,267 +884,244 @@ def LSTMCellS(x, hx, cx, w_ih, w_hh, b_ih, b_hh):
-  %13 : Tensor = aten::add(%10, %12, %8)
-  %14 : Tensor = aten::add(%13, %b_ih, %8)
-  %gates : Tensor = aten::add(%14, %b_hh, %8)
-  %16 : Tensor[] = aten::chunk(%gates, %7, %8)
-  %ingate.1 : Tensor, %forgetgate.1 : Tensor, %cellgate.1 : Tensor, %outgate.1 : Tensor = prim::ListUnpack(%16)
-  %ingate : Tensor = aten::sigmoid(%ingate.1)
-  %forgetgate : Tensor = aten::sigmoid(%forgetgate.1)
-  %cellgate : Tensor = aten::tanh(%cellgate.1)
-  %outgate : Tensor = aten::sigmoid(%outgate.1)
-  %25 : Tensor = aten::mul(%forgetgate, %cx)
-  %26 : Tensor = aten::mul(%ingate, %cellgate)
-  %cy : Tensor = aten::add(%25, %26, %8)
-  %28 : Tensor = aten::tanh(%cy)
-  %hy : Tensor = aten::mul(%outgate, %28)
-  %30 : (Tensor, Tensor) = prim::TupleConstruct(%hy, %cy)
-  return (%30)
-```
-
-Execution starts in `GraphExecutor::run`, which takes a Stack of inputs.
-
-*Specialization* The executor *specializes* the Graph for the particular set of inputs. Specialization is handled by the `ArgumentSpec` object which extracts a "signature" composed of all the properties being specialized. We only specialize to the properties of Tensors. The ArgumentSpec only records properties for Tensors that either appear directly in the inputs to the graph or inside Tuples that are inputs to the Graph. The properties recorded are currently:
-
-* dtype
-* rank, but not size
-* requires_grad
-* device type (cpu, cuda)
-* defined - whether the Tensor exists or is a placeholder
+### Profiling ###
 
-The ArgumentSpec object is used as a key into a cache that holds pre-optimized Code objects (held in an ExecutionPlan object). On a cache hit, an InterpreterState is created and the Code in the cache is run.
-
-
-*Pre-derivative Optimization* On a code cache miss, we generate a new optimized Graph on the fly (`compileSpec`). It starts by creating a copy of the initial Graph and setting the input types to the specialized Tensor types observed in this specialization. TensorType inputs to the Graph will get refined with types that know the device, number of dimensions, and requires grad state.
+In the profiling stage, a Profiling Graph Executor takes an unoptimized graph runs required and passes that doesn't need profiling information ([see the next section](#Optimization)) and inserts `prim::profile` on **tensor-typed** inputs to fusible operations. 
+`torch::jit::ProfilingRecord` also maintains a mapping between output values of `prim::profile` nodes and the properties of input tensors at runtime.
+This mapping is maintained for each invocation of a function identified by Interpreter's `frame_id` (which is globally unique).
+It's important to not mix the tensor properties from different invocations for the same `torch::jit::Value*`, as this may inhibit symbolic (dynamic) shape inference to correctly inferring the relationships between the arguments of an operation and its outputs.                                                                                     
```

Comment:
A bit unclear: it's important to whom? Is it important for the profiling executor (and we take care of that by doing this and that), or is it important for users of the profiling executor and they need to be aware of something?